### PR TITLE
[windows] redo windows tailer as polling rather than async.

### PR DESCRIPTION
**Pull request by derekwbrown**
### What does this PR do?

adds support for logs (windows)

### CODE REVIEW

File: pkg/logs/input/tailer/tailer.go [hunk 1]
```diff
@@ -45,7 +42,7 @@ type Tailer struct {
 	closeTimeout time.Duration
 	shouldStop   bool
 	stopTimer    *time.Timer
-	stopMutex    sync.Mutex
+	stopMutex    sync.Mutex;
```
Comment (tmichelet) [inline pkg/logs/input/tailer/tailer.go]: nit: `;`

File: pkg/logs/input/tailer/tailer_nix.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	if inode(stat1) != inode(stat2) {
+		return true, nil
+	}
+
+	if stat1.Size() < t.GetReadOffset() {
+		return true, nil
+	}
+	return false, nil
+}
```
Comment (tmichelet) [inline pkg/logs/input/tailer/tailer_nix.go]: nit: no new line

File: pkg/logs/input/tailer/tailer_nix.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+// +build !windows
```
Comment (tmichelet) [inline pkg/logs/input/tailer/tailer_nix.go]: agent doesn't compile on unix:
```
# github.com/DataDog/datadog-agent/pkg/logs/input/tailer
pkg/logs/input/tailer/tailer_nix.go:11:2: imported and not used: "fmt"
pkg/logs/input/tailer/tailer_nix.go:15:2: imported and not used: "sync"
pkg/logs/input/tailer/tailer_nix.go:16:2: imported and not used: "sync/atomic"
pkg/logs/input/tailer/tailer_nix.go:17:2: imported and not used: "time"
pkg/logs/input/tailer/tailer_nix.go:21:2: imported and not used: "github.com/DataDog/datadog-agent/pkg/logs/config"
pkg/logs/input/tailer/tailer_nix.go:23:2: imported and not used: "github.com/DataDog/datadog-agent/pkg/logs/message"
pkg/logs/input/tailer/tailer_nix.go:93:5: undefined: inode
```

Happy to fix it myself if you don't have a working unix environment

File: pkg/logs/input/tailer/tailer_nix.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		if n == 0 {
+			t.wait()
+			continue
+		}
+		t.d.InputChan <- decoder.NewInput(inBuf[:n])
+		t.incrementReadOffset(n)
+	}
+}
+func (t *Tailer) checkForRotation() (bool, error) {
+	f, err := os.Open(t.path)
```
Comment (tmichelet) [inline pkg/logs/input/tailer/tailer_nix.go]: this method has been updated on the master branch, see https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/tailer/scanner.go#L151-L170

Could you port the updates here?

**Review by tmichelet — COMMENTED** [review]
Thanks Derek, works as expected with little forking from unix!

Could you address the small points, remove the debug logs and comments?

Also, I think we can make `checkForRotation` very close on windows and on unix, so let's try to implement that too

**Review by tmichelet — COMMENTED** [review]
Getting close! Added a bunch of small comments on this part of the code.

Also, one feature we have is storing the offset of the last log line we tailed and starting again from there when the agent restarts. This currently assumes that a line ends with `\n`, which is not exactly true on windows. A quick fix is to add a EOL_SIZE constant and update these two lines:
- https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/decoder/line_handler.go#L81
- https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/decoder/line_buffer.go#L39

Could you have a look at that?

~Lastly, when testing this I noticed a bug in the logic to store the offset, when the log line ends with a whitespace we trim it and store the wrong offset. Let's not fix this right now but keep that in mind in case you end up in this edge case in your own testing.~ This bug is fixed by #1095

**Comment by tmichelet** [comment]
Update: I was wrong in my comment, there was indeed an issue but it was unrelated to the size of EOL, and is fixed by last commit

**Review by ajacquemot — CHANGES_REQUESTED** [review]

File: pkg/logs/input/tailer/tailer_windows.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	} else {
+		log.Debugf("Error stat()ing file %v", err)
+	}
+	f.Seek(t.GetReadOffset(), io.SeekStart)
+	inBuf := make([]byte, 4096)
+
+	for {
+		n, err := f.Read(inBuf)
+		if n == 0 || err != nil {
```
Comment (ajacquemot) [inline pkg/logs/input/tailer/tailer_windows.go]: I think you should return the error to catch `EOF`

File: pkg/logs/input/tailer/tailer_windows.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+			log.Debug("File size now zero, resetting offset")
+			t.SetReadOffset(0)
+			t.SetDecodedOffset(0)
+		} else if sz < t.GetReadOffset() {
+			log.Debug("Offset off end of file, resetting")
+			t.SetReadOffset(0)
+			t.SetDecodedOffset(0)
+		}
+	} else {
+		log.Debugf("Error stat()ing file %v", err)
```
Comment (ajacquemot) [inline pkg/logs/input/tailer/tailer_windows.go]: What does it mean to get an error on file stats ? Don't we want to return this error ?

Comment (tmichelet) [reply pkg/logs/input/tailer/tailer_windows.go]: Not sure, but it's not 100% required for the feature so we'll go ahead

File: pkg/logs/input/tailer/tailer_windows.go (hunk 5 above)
Comment (tmichelet) [reply pkg/logs/input/tailer/tailer_windows.go]: good point

**Pull request merged**